tests/pos/sams.scala

```diff
@@ -0,0 +1,75 @@
+object test {
+
+  trait X { def foo(x: Int): Int; def bar = foo(2) }
+
+  val x: X = (x: Int) => 2  // should be a closure
+
+  trait T {
+    var f = 2
+    def foo(x: Int): Int
+  }
+
+  val t: T = (x: Int) => 2   // needs to be an anonymous class because of defined field
+
+  trait U extends T
+
+  val u: U = (x: Int) => 2   // needs to be an anonymous class because of inherited field
+
+  trait V extends Exception { def foo(x: Int): Int }
+
+  val v: V = (x: Int) => 2   // needs to be an anonymous class because the trait extends a non-object class
```

Comment:
> we expect that lambda.isInstanceOf[C]. 

This is the case in this example:
```scala
object Foo {
  trait V extends Exception { def foo(x: Int): Int }

  val v: V = (x: Int) => 2

  def main(args: Array[String]): Unit = {
    println(v.isInstanceOf[V])
  }
}
```
In master right now it will print "true". This makes sense: for each trait, Dotty will generate an interface, and `LambdaMetaFactory` will return an instance of that interface.

@DarkDimius : Your example is something else entirely: abstract class should not be SAM types, only traits can be SAM types.